server/app/survey_response.go

```diff
@@ -160,3 +231,109 @@ func (a *UserSurveyApp) UpdatePostForExpiredSurvey(userID, surveyID string) erro
 
 	return nil
 }
+
+func (a *UserSurveyApp) updateNPSScoreGroupCount(survey *model.Survey, oldResponse, newResponse *model.SurveyResponse) error {
+	// survey table has a column each for number of promoters, neutral and detractors.
+	// Depending on the new score, we need to increment the corresponding columns, and
+	// depending on the old score, decrement its corresponding column value.
+	//
+	// Since we store ratings as soon as user selects it, and then we update the whole response
+	// when user clicks the "submit" button, it is possible to have an existing response saved
+	// in the database already.
+	//
+	// This function computes three values, either 1 or -1, one for each promoter, neutral and detractors columns.
+	// We then add the computed values corresponding to each column in the database, effectively
+	// increasing the column corresponding the new score by 1 and incrementing the column
+	// corresponding to the old score by -1 (incrementing by -1 is same as decrementing by 1, incrementing by -1 here to keep SQL queries simple).
+
+	systemRatingQuestionID, err := survey.GetSystemRatingQuestionID()
+	if err != nil {
+		return errors.Wrap(err, "updateNPSScoreGroupCount: failed to find a system rating question in survey")
+	}
+
+	var oldPromoterFactor, oldNeutralFactor, oldDetractorFactor int
+	var newPromoterFactor, newNeutralFactor, newDetractorFactor int
+
+	newRating, err := strconv.Atoi(newResponse.Response[systemRatingQuestionID])
+	if err != nil {
+		a.api.LogError("updateNPSScoreGroupCount: failed to convert new rating value from string to number")
+		return errors.Wrap(err, "updateNPSScoreGroupCount: failed to convert new rating value from string to number")
+	}
+
+	newPromoterFactor, newNeutralFactor, newDetractorFactor = a.getRatingGroupFactors(newRating)
+
+	if oldResponse != nil {
+		oldRating, err := strconv.Atoi(oldResponse.Response[systemRatingQuestionID])
+		if err != nil {
+			a.api.LogError("updateNPSScoreGroupCount: failed to convert new rating value from string to number")
+			return errors.Wrap(err, "updateNPSScoreGroupCount: failed to convert new rating value from string to number")
+		}
+
+		if !a.hasRatingGroupChanged(oldRating, newRating) {
+			// nothing to do if rating groups are unchanged
+			return nil
+		}
+
+		oldPromoterFactor, oldNeutralFactor, oldDetractorFactor = a.getRatingGroupFactors(oldRating)
+
+		oldPromoterFactor *= -1
+		oldNeutralFactor *= -1
+		oldDetractorFactor *= -1
+	}
+
+	var promoterFactor, neutralFactor, detractorFactor int
+
+	if oldPromoterFactor != 0 {
```

Comment:
Nit: could separate this into a function to avoid repetition

Reply:
Done. Added a `CoalesceInt` function to get first non 0 number from a list of number. The new code look like this-

```go
promoterFactor := utils.CoalesceInt(oldPromoterFactor, newPromoterFactor)
neutralFactor := utils.CoalesceInt(oldNeutralFactor, newNeutralFactor)
detractorFactor := utils.CoalesceInt(oldDetractorFactor, newDetractorFactor)
```